### ISSUE TITLE
Module Namespacing Support for Packages

### ISSUE THREAD

**Issue by krpeacock** [expo/snack#161]
### Summary

When installing a package that uses module namespacing (`export * as {moduleName} from './{moduleName}`), Expo's default configuration throws an error when compiling for web and iOS. This is an important tool for structuring TypeScript code, so I'd expect it to work out of the box.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

iOS, Web

### SDK Version (managed workflow only)

4.5.2

### Environment

MacOS, Node v14.15.5.

### Reproducible demo or steps to reproduce from a blank project

https://snack.expo.io/kSlfJiSnR

Steps to reproduce:
- `npm install @dfinity/agent`
- Somewhere in the code - `import { HttpAgent } from "@dfinity/agent";`

**Comment by krpeacock** [comment]
Thanks for the thorough investigation! You're totally correct on 1 and 3 - I stopped adding to the example once I could reproduce the error I was aware of. 

4 is another known issue, but we'll avoid shipping a TextEncoder polyfill due to its wide browser support and that there are good options available on npm that a RN dev could include.

As for 2, I think that's the heart of the issue, and I don't think it's a Snack bug. I first reproduced the issue on my dev environment using the Expo CLI. Here's a more complete example to test locally: https://github.com/krpeacock/dfinity-expo-proof-of-concept

**Comment by byCedric** [comment]
Thanks for that! I can't seem to get the repo working, it's missing this file:
> `./.dfx/local/canisters/counter/counter.js`

Is that correct?

**Comment by krpeacock** [comment]
Ah right, that's the domain-specific stuff. I'll unignore that folder and push it up for reproduction to spare you from needing to install our SDK. You won't end up with a fully working example, but the frontend should compile

**Comment by krpeacock** [comment]
Should be good to go!

If you wanted to test the full stack example, the commands to set up would be 
```
sh -ci "$(curl -fsSL https://sdk.dfinity.org/install.sh)"
dfx start --background
dfx deploy
npm start
```

**Comment by byCedric** [comment]
Awesome, thanks! When I add the following libraries:
```
$ npm install --save buffer text-encoding-polyfill
```

And add this to the top of the `App.tsx` file:

```
// Add TextEncoder polyfill
import 'text-encoding-polyfill';
// Add Buffer polyfill
import { Buffer } from 'buffer';
global.Buffer = Buffer;
```

It seems to work, without compiler issues. Only the warning that it can't connect to the network, which makes sense since I don't have that. 😄

**Comment by krpeacock** [comment]
That's super interesting! I'm still getting the `export` error on my machine when I try to run the web build. Were you compiling to iOS or Android?

![image](https://user-images.githubusercontent.com/16298804/121716939-7f52c200-ca95-11eb-9b87-bec59e944b7d.png)

**Comment by byCedric** [comment]
Ah, you are compiling to web! For me, it works on both Android and iOS. Haven't tested web, but I know what the problem could be. On web, we are using a plain webpack configuration to build the app. It does not transpile `node_modules`, because we expect modules for the browser to be pre-compiled for the browser. I don't think the majority of web browsers support the new module syntax, which means you have to manually add the module to be transpiled.

To do this, you have to run `$ expo customize:web` and [change the webpack config to this](https://github.com/expo/expo-cli/tree/master/packages/webpack-config#include-modules):

```js
const createExpoWebpackConfigAsync = require('@expo/webpack-config');

module.exports = async function (env, argv) {
  const config = await createExpoWebpackConfigAsync(
    {
      ...env,
      babel: {
        dangerouslyAddModulePathsToTranspile: [
          // Ensure that all packages starting with @dfinity are transpiled.
          '@dfinity',
        ],
      },
    },
    argv
  );
  return config;
};
```

You might be better of including a pre-transpiled copy of the code, or have users add this configuration.

**Comment by krpeacock** [comment]
Great, this sounds viable then with some documentation updates!

As an ideal solution, Is there a way we could build a pre-transpiled version in addition to the harmony exports and signal that to `Expo` via the `package.json`? Or would you recommend we create an additional pre-transpiled package to publish to `npm`?

**Comment by byCedric** [comment]
Yes, that's also possible! I'm not sure how Snack would react, but we can always find a solution for that later.

  - With normal npm modules, you should be able to define [`main`](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#main) file as entrypoint (defaults to `index.js`).
  - With tools like Rollup and webpack, you can define a [`module`](https://github.com/rollup/rollup/wiki/pkg.module) entrypoint. (although this isn't an "official" spec, it's still in proposal state)
  - With React Native, you can define a [`react-native`](https://github.com/callstack/react-native-paper/blob/main/package.json#L7) entrypoint, e.g. including the polyfills. (couldn't find the docs for this, but you can see it there in action)

All of these can be defined in the `package.json`. Again, I'm not sure how Snack would react to these entry points but we can find something for that later on.
